README.md
# AutoTrie

A versatile library which solves autocompletion in Dart/Flutter. It is based around
a space-efficient implementation of Trie which uses variable-length lists. With this, serving
auto-suggestions is both fast and no-hassle. Suggestions are also sorted by how often 
they've been entered and subsorted by recency of entry, for search-engine-like results.

Read more about Trie [here][trie].

[trie]: https://medium.com/basecs/trying-to-understand-tries-3ec6bede0014

## A Brief Note

It takes time, effort, and mental power to keep this package updated, useful, and
improving. If you used or are using the package, I'd appreciate it if you could spare a few 
dollars to help me continue development.

[![PayPal](https://img.shields.io/static/v1?label=PayPal&message=Donate&color=blue&logo=paypal&style=for-the-badge&labelColor=black)](https://www.paypal.me/kishoredev)

## Usage

A usage example is provided below. Check the API Reference for detailed docs:

```dart
import 'package:autotrie/autotrie.dart';

void main() {
  // You are allowed to initialize with a starting databank by passing a `bank` parameter.
  var engine = AutoComplete(engine: SortEngine.configMulti(Duration(seconds: 1), 15, 0.5, 0.5));

  engine.enter('more'); // Enter more thrice.
  engine.enter('more');
  engine.enter('more');

  engine.enter('moody'); // Enter moody twice.
  engine.enter('moody');

  engine.enter('morose'); // Enter scattered words (with mo).
  engine.enter('morty');
  engine.enter('moment');
  engine.enter('momentum');

  engine.enter('sorose'); // Enter scattered words (without mo).
  engine.enter('sorty');

  engine.delete('morose'); // Delete morose.

  // Check if morose is deleted.
  print('Morose deletion check: ${!engine.contains('morose')}');

  // Check if engine is empty.
  print('Engine emptiness check: ${engine.isEmpty}');
 
  // They've been ranked by frequency and recency. Since they're all so similar
  // in recency, frequency takes priority.
  print("'mo' suggestions: ${engine.suggest('mo')}");
  // Result: [more, moody, momentum, moment, morty]

  // Get all entries.
  // They've *not* been sorted.
  // Use `engine.suggest('')` to get all entries sorted` 
  print('All entries: ${engine.allEntries}');
  // Result: [more, moody, sorty, sorose, momentum, moment, morty]
}

```

## Sorting
The AutoComplete constructor takes a SortEngine, which it uses to sort the result of the autocompletion operation.
There are a few different modes it can operate in:

* SortEngine.entriesOnly() -> AutoComplete results are only sorted by number of entries in the engine (High to Low)
* SortEngine.msOnly() -> AutoComplete results are only sorted by how much time has passed since their last entry (Low to High)
* SortEngine.simpleMulti() ->
    - Sorted using two logistic curves, one for ms and one for entries
        * The ms curve is set to use 3 years (a LOT) as the upper end of how far back entries could have been entered
        * The entries curve is set to use 30 entries as the max amount of entries
        * These values are highly arbitrary and not likely to fit your project; this mode is **not** recommended unless
        you are just playing around.
    - Takes two weights (one for recency and one for entries) which can be used to balance how heavily each factor
    should affect the final sorting.
* SortEngine.configMulti() ->
    - Sorted using two logistic curves, one for ms and one for entries
        * The ms curve is balanced using a parameter (a Duration) for the max time since entry in this engine.
        * The entries curve is balanced using a parameter (an int) for the max amount of entries in this engine.
        * If you know approximately how old and how big this AutoComplete engine is, it is **highly** recommended that
        you use this mode.
    - Takes two weights (one for recency and one for entries) which can be used to balance how heavily each factor
    should affect the final sorting.
    
## Basic File Persistence
AutoComplete is natively capable of writing itself to and reading itself from a file. To do this, persist to a file
using the `persist` method (it takes a `File` object):
`await engine.persist(myFile);`

Then you can rebuild using `AutoComplete.fromFile` (it takes a `File` along with the mandatory `SortEngine`):
`var engine = AutoComplete.fromFile(file: myFile, engine: SortEngine.entriesOnly());`

This persistence will preserve all the metadata (last insert, number of entries) in the table as well as the 
core data (the Strings themselves).

## Hive Integration
- [Hive][hive] is a speedy, local, and key-value database for Dart/Flutter. Go check it out if you haven't already!
- Hive integration is now available with autotrie:
    - Our way of integration uses extension methods.
    - Import Hive and AutoTrie, and create a normal Hive box using `Hive.openBox('nameHere')`.
    - You can then call `searchKeys(String prefix)` and `searchValues(String prefix)` on that box to get auto-suggestions.
    - There is no sorting options: only entry-level sorting is available.

[hive]: https://pub.dev/packages/hive

## Features and bugs

Please file feature requests and bugs at the [issue tracker][tracker].

[tracker]: https://github.com/AKushWarrior/autotrie/issues

---
###### This library and its contents are subject to the terms of the Mozilla Public License, v. 2.0.
###### © 2020 Aditya Kishore

CHANGELOG.md
## 1.0.0 - 1.0.0+1
- Added complex sorting support
- Added batteries-included file persistence
- Updated dependencies
- Increased efficiency due to elimination of double sorting (one-step instead)
- Updated README

## 0.4.1
- Fixed Hive integration (was defunct)
- It now uses extension methods

## 0.3.1
- Added subsorting by recency
- Removed isEnd to save space and replaced it with > 0 checks
- Somewhat better/more detailed docs

## 0.2.1
- Added integration with Hive
- Better documentation

## 0.1.2
- Added methods for adding List<String>
- Better documentation

## 0.1.1 - 0.1.1+1
- Initial commit and publish
- Provides Trie engine for autocompletion
- Has all the basic methods


test/autotrie_test.dart
import 'package:test/test.dart';

void main() {
  group('A group of tests', () {

    setUp(() {
    });

    test('First Test', () {
      expect(true, isTrue);
    });
  });
}


lib/src/datatree/datatree_test.dart
import 'dart:io';

import 'package:autotrie/autotrie.dart';
import 'tree.dart';

void main() {
  var tree = TrieSearchTree(SortEngine.configMulti(Duration(seconds:30), 5, 0.25, 0.75).scoreFunc);

  tree.addWord('more'); //enter more thrice
  tree.addWord('more');
  tree.addWord('more');
  sleep(Duration(seconds: 1));

  tree.addWord('moody'); //enter moody twice
  tree.addWord('moody');
  sleep(Duration(seconds: 1));

  tree.addWord('morose'); //enter scattered words (with mo)
  sleep(Duration(seconds: 1));
  tree.addWord('morty');
  sleep(Duration(seconds: 1));
  tree.addWord('moment');
  sleep(Duration(seconds: 1));
  tree.addWord('momentum');
  sleep(Duration(seconds: 1));

  tree.addWord('sorose'); //enter scattered words (without mo)
  sleep(Duration(seconds: 1));
  tree.addWord('sorty');
  sleep(Duration(seconds: 1));

  //tree.remove('morose');

  print(tree.suggestions('')); // [more, moody, morty, moment, momentum]
}

lib/src/autotrie_base.dart
import 'dart:io';
import 'dart:math';

import 'package:hive/hive.dart';
import 'package:meta/meta.dart';

part 'datatree/tree.dart';

/// Engine for auto-completion of Strings.
///
/// Add entries to build up the suggestion bank. Then, you can use the suggest
/// method to get the auto-completions for the beginning of a String.
class AutoComplete {
  _TrieSearchTree _tree;

  /// Whether this autocompletion engine has changed since the last time
  /// `persist` was called.
  bool changedSincePersist = true;

  /// Constructs an instance of `AutoComplete`.
  ///
  /// This uses [engine] to sort the results of autocompletion.
  ///
  /// If you pass a [bank] parameter, the engine will have all the Strings in
  /// bank as search results.
  ///
  /// If you enter a word multiple times, it will be prioritized in the search
  /// results, since search results are sorted by number of times entered.
  ///
  /// If multiple words have the same number of entries, they are sorted by recency,
  /// with the most recently entered word being on top.
  AutoComplete({@required SortEngine engine, List<String> bank}) {
    _tree = _TrieSearchTree(engine.scoreFunc);
    bank ??= <String>[];
    for (var x in bank) {
      enter(x);
    }
  }

  AutoComplete.fromFile({@required File file, @required SortEngine engine}) {
    var fileBank = file.readAsLinesSync();
    var tree = _TrieSearchTree(engine.scoreFunc);
    for (var x in fileBank) {
      var values = x.split(' | ');
      tree.addWordWithParams(values[0], int.parse(values[1]), int.parse(values[2]));
    }
    _tree = tree;
  }

  /// Add an entry to the engine.
  ///
  /// The engine will now include `entry' as a search result. If you enter a word
  /// multiple times, it will be prioritized in the search results, since search
  /// results are sorted by number of times entered.
  ///
  /// If multiple words have the same number of entries, they are sorted by recency,
  /// with the most recently entered word being first.
  void enter(String entry) {
    _tree.addWord(entry);
    changedSincePersist = true;
  }

  /// Add multiple entries to the engine.
  ///
  /// The engine will now include the contents of `entries` as a search result.
  /// If you enter a word multiple times, it will be prioritized in the search
  /// results, since search results are sorted by number of times entered.
  ///
  /// If multiple words have the same number of entries, they are sorted by recency,
  /// with the most recently entered word being on top.
  void enterList(List<String> entries) {
    for (var x in entries) {
      enter(x);
    }
    changedSincePersist = true;
  }

  /// Clear all the entries. The engine is now blank.
  void clearEntries() {
    _tree.root = TrieNode('', false);
    changedSincePersist = true;
  }

  /// Get all the entries in a list.
  ///
  /// This is NOT sorted. Use [suggest('')] to get all results, sorted.
  Iterable<String> get allEntries => _tree.all.map((e) => e.value);

  /// Suggest entries based on the beginning of the string.
  ///
  /// This method returns a List<String>, which contains the suggestions (entries).
  /// These suggestions are ordered by the number of times the suggestion has been
  /// entered.
  ///
  /// If multiple suggestions have the same number of entries, they are sorted by recency,
  /// with the most recently entered suggestion being on top.
  List<String> suggest(String prefix) {
    return _tree.suggestions(prefix);
  }

  /// Returns true if this engine has no entries.
  bool get isEmpty => _tree.root.children.isEmpty;

  /// Returns true if this engine contains `entry`.
  bool contains(String entry) => _tree.search(entry);

  /// Deletes `entry` from the engine if it exists.
  ///
  /// If `entry` is not in the engine, it will do nothing.
  void delete(String entry) => _tree.remove(entry);

  /// Takes a file and writes the entire engine to it. You can then rebuild this
  /// engine at some point in the future from the file.
  Future<void> persist(File file) async {
    var entriesToWrite = _tree.all;
    var sink = file.openWrite();
    sink.writeAll(entriesToWrite, '\n');
    await sink.flush();
    await sink.close();
  }
}

extension AutoCompleteBox on Box {
  /// Gives suggested auto-complete keys from this box, along with corresponding
  /// values. Suggested values are sorted by number of occurrences in this box.
  ///
  /// If all keys are not String, this will instead call the toString() method
  /// of non-String keys and search the results.
  ///
  /// Suggestions are returned in a Map, where:
  /// - The keys are the autocomplete suggestions.
  /// - The values are the corresponding values from this box.
  Map<dynamic, dynamic> searchKeys(String keyPrefix) {
    var keyList = keys.map((e) => e.toString()).toList();
    var _engineKeys =
        AutoComplete(engine: SortEngine.entriesOnly(), bank: keyList);
    var keysuggest = _engineKeys.suggest(keyPrefix);
    var map = toMap();
    map.removeWhere((dynamic key, dynamic val) {
      return !keysuggest.contains(key.toString());
    });
    return map;
  }

  /// Gives suggested auto-complete values from this box, along with corresponding
  /// keys. Suggested values are sorted by number of occurrences in this box.
  ///
  /// If all keys are not String, this will instead call the toString() method
  /// of non-String keys and search the results.
  ///
  /// Suggestions are returned in a Map, where:
  /// - The keys are the correspondent keys to the autocomplete suggestions.
  /// - The values are the autocomplete suggestions.
  Map<dynamic, dynamic> searchValues(String valuePrefix) {
    var valList = values.map((e) => e.toString()).toList();
    var _engineValues =
        AutoComplete(engine: SortEngine.entriesOnly(), bank: valList);
    var valuesuggest = _engineValues.suggest(valuePrefix);
    var map = toMap();
    map.removeWhere((dynamic key, dynamic val) {
      return !valuesuggest.contains(val.toString());
    });
    return map;
  }
}

class SortEngine {
  double Function(SortValue element) scoreFunc;

  /// You can manually define a function to score a given [_SortValue] on a
  /// scale of 0 to 1. A [_SortValue] consists of [msToNow], which is how many
  /// ms ago the entry was submitted, and [numEntries], how many times the entry
  /// has been submitted.
  SortEngine.manual(this.scoreFunc);

  /// A simple engine that sorts entries based on how recently they've been
  /// entered. This engine does not use any number-of-entries sorting.
  SortEngine.msOnly() {
    scoreFunc = (SortValue a) => -(a.msToNow.toDouble());
  }

  /// A simple engine that sorts entries based on how many times they've been
  /// entered. This engine does not use any recency sorting.
  SortEngine.entriesOnly() {
    scoreFunc = (SortValue a) => a.numEntries.toDouble();
  }

  /// A complex multi sorting engine with user-defined value curves.
  ///
  /// It sets [maxEntries] as the maximum number of entries; beyond that point, all
  /// numbers of entries will have the same (max) prioritization. You should define
  /// this as the maximum number of repeated entries you think is likely for your
  /// situation.
  ///
  /// It also sets [timeScale] as the maximum time since entry; beyond that point, all
  /// time since entries will have the same (min) prioritization. You should define
  /// this as the oldest entry you think is likely for your situation.
  ///
  /// Both curves follow a logistic pattern, with an exponential period lasting
  /// until their respective max points. The entries curve is traditional (more
  /// entries --> better score) and the time curve is flipped (more time --> worse
  /// score). As noted above, both max points are user defined.
  SortEngine.configMulti(
      Duration timeScale, int maxEntries, double msWeight, double entryWeight) {
    if (msWeight + entryWeight != 1.0) {
      throw ArgumentError('msWeight + entryWeight must equal 1.');
    }
    scoreFunc = (SortValue a) {
      var msScore =
          1 / pow(1.5, (1 / timeScale.inMilliseconds) * a.msToNow) * msWeight;
      var entryScore =
          (1 + (-1 / pow(1.5, (1 / maxEntries) * a.numEntries))) * entryWeight;
      return entryScore + msScore;
    };
  }

  /// A simpler multi sorting engine with predefined value curves.
  ///
  /// It sets 30 as the maximum number of entries; beyond that point, all
  /// numbers of entries will have the same (max) prioritization.
  ///
  /// It also sets ~3 years as the maximum time since entry; beyond that point, all
  /// time since entries will have the same (min) prioritization.
  ///
  /// Both curves follow a logistic pattern, with an exponential period lasting
  /// until their respective max points. The entries curve is traditional (more
  /// entries --> better score) and the time curve is flipped (more time --> worse
  /// score). As noted above, both max points are user defined.
  ///
  /// If you think that you will have significantly more or less than 30 max entries,
  /// or significantly more or less than ~3 years max time since entry, you are
  /// encouraged to use [configMulti].
  SortEngine.simpleMulti(double msWeight, double entryWeight) {
    if (msWeight + entryWeight != 1.0) {
      throw ArgumentError('msWeight + entryWeight must equal 1.');
    }
    scoreFunc = (SortValue a) {
      var msScore = 1 / pow(1.5, (1 / 0x174876E800) * (a.msToNow)) * msWeight;
      var entryScore =
          (1 + (-1 / pow(1.5, (1 / 30) * a.numEntries))) * entryWeight;
      return entryScore + msScore;
    };
  }
}

class SortValue {
  final int _msSinceEpoch;

  int get msToNow => DateTime.now().millisecondsSinceEpoch - _msSinceEpoch;
  final int numEntries;

  SortValue(this._msSinceEpoch, this.numEntries);
}


lib/src/datatree/tree.dart
part of '../autotrie_base.dart';

class _TrieSearchTree {
  TrieNode root;
  double Function(SortValue e) sort;

  _TrieSearchTree (this.sort) {
    root = TrieNode('', false);
    root.children = <TrieNode>[];
  }

  void addWord (String word) {
    var base = root;

    //Iterate through string and add/progress through nodes.
    for (var i = 0; i<word.length; i++) {
      var x = TrieNode(word[i], false);
      if (!base.children.contains(x)) {
        // x is added to base.children
        base.children.add(x);
      } else {
        // x points to base.children version
        x = base.children.where((e) => e==x).first;
      }
      if (i == word.length-1) {
        x.lastInsert = DateTime.now().millisecondsSinceEpoch;
        x.hits++;
      }
      base = x;
    }
  }

  void addWordWithParams (String word, int hits, int lastInsert) {
    var base = root;

    //Iterate through string and add/progress through nodes.
    for (var i = 0; i<word.length; i++) {
      var x = TrieNode(word[i], false);
      if (!base.children.contains(x)) {
        // x is added to base.children
        base.children.add(x);
      } else {
        // x points to base.children version
        x = base.children.where((e) => e==x).first;
      }
      if (i == word.length-1) {
        x.lastInsert = lastInsert;
        x.hits = hits;
      }
      base = x;
    }
  }

  List<TrieString> get all {
    var base = root;
    if (base.children == <TrieNode>[] && base.hits > 0) return [];
    var returner = <TrieString>[];
    _suggestRec(base, '', returner);
    return returner;
  }

  List<String> suggestions (String prefix) {
    var base = root;
    for (var i = 0; i<prefix.length; i++) {
      var x = TrieNode(prefix[i], false);
      if (base.children.contains(x)) {
        base = base.children[base.children.indexOf(x)];
      } else {
        return [];
      }
    }
    if (base.children == <TrieNode>[] && base.hits > 0) return [prefix];
    var returner = <TrieString>[];
    _suggestRec(base, prefix, returner);

    returner.sort((TrieString a, TrieString b) {
      var sortA = sort(SortValue(a.lastInsert, a.hits));
      var sortB = sort(SortValue(b.lastInsert, b.hits));

      if (sortA < sortB) {
        return 1;
      } else if (sortA == sortB) {
        return 0;
      } else {
        return -1;
      }
    });
    return returner.map((e) => e.value).toList();
  }

  void _suggestRec (TrieNode node, String word, List<TrieString> returner) {
    if (node.hits > 0) returner.add(TrieString(word, node.hits, node.lastInsert));
    for (var n in node.children) {
      _suggestRec(n, word + n.value, returner);
    }
  }

  bool search (String word) {
    var base = root;
    for (var i = 0; i < word.length; i++) {
      var x = TrieNode(word[i], false);
      if (!base.children.contains(x)) return false;
      base = base.children[base.children.indexOf(x)];
    }
    return true;
  }

  void remove (String word) {
    if (root.children.isEmpty) {
      return;
    } else if (!search(word)) {
      return;
    }
    _delete(root, word, 0);
  }

  bool _delete (TrieNode cur, String word, int index) {
    if (index == word.length) {
      return cur.children.isEmpty;
    }
    var ch = word[index];
    var next = cur.children[cur.children.indexOf(TrieNode(ch, index == word.length-1))];
    if (next == null) {
      return false;
    }
    var shouldDeleteNode = _delete(next, word, index + 1) && (!(next.hits>0) || index == word.length -1);

    if (shouldDeleteNode) {
      cur.children.remove(next);
      return cur.children.isEmpty;
    }
    return false;
  }
}

class TrieString {
  String value;
  int hits;
  int lastInsert;

  TrieString(String value, int hits, int insert) {
    this.value = value;
    this.hits = hits;
    lastInsert = insert;
  }

  @override
  String toString() {
    return '$value | $hits | $lastInsert';
  }
}

class TrieNode {
  String value;
  List<TrieNode> children;
  int lastInsert;
  int hits = 0;

  TrieNode(String value, bool isEnd) {
    this.value = value;
    children = <TrieNode>[];
  }

  @override
  bool operator ==(other) {
    return value == other.value;
  }
}

example/autotrie_example.dart
import 'package:autotrie/autotrie.dart';
import 'package:autotrie/src/autotrie_base.dart';

void main() {
  var engine = AutoComplete(engine: SortEngine.configMulti(Duration(seconds: 1), 15, 0.5, 0.5)); //You can also initialize with a starting databank.

  engine.enter('more'); // Enter more thrice.
  engine.enter('more');
  engine.enter('more');

  engine.enter('moody'); // Enter moody twice.
  engine.enter('moody');

  engine.enter('morose'); // Enter scattered words (with mo).
  engine.enter('morty');
  engine.enter('moment');
  engine.enter('momentum');

  engine.enter('sorose'); // Enter scattered words (without mo).
  engine.enter('sorty');

  engine.delete('morose'); // Delete morose.

  // Check if morose is deleted.
  print('Morose deletion check: ${!engine.contains('morose')}');

  // Check if engine is empty.
  print('Engine emptiness check: ${engine.isEmpty}');

  // Suggestions starting with 'mo'.
  // They've been ranked by frequency and recency. Since they're all so similar
  // in recency, frequency takes priority.
  print("'mo' suggestions: ${engine.suggest('mo')}");
  // Result: [more, moody, momentum, moment, morty]

  // Get all entries.
  // They've not been sorted.
  print('All entries: ${engine.allEntries}');
  // Result: [more, moody, sorty, sorose, momentum, moment, morty]
}

// Check the API Reference for the latest information and adv.
// methods from this class.